### PR TITLE
Remotenames

### PR DESCRIPTION
**Pull request by skyfrench**
I have made some changes with the aim of introducing naming conventions for git remote repositories. 

dls-checkout-module.py will now check for the presence of a .gitremotes file with the following format (example from ADEiger):

dls-controls git@github.com:dls-controls/ADEiger.git
gitolite ssh://dascgitolite@dasc-git.diamond.ac.uk/controls/support/ADEiger
upstream git@github.com:areaDetector/ADEiger.git

If this file is present it will automatically add the listed remotes.

In addition - when a repository is cloned from gitolite the repository will be cloned with the name gitolite as default.

### CODE REVIEW

**Comment by willrogers** [comment]
Looks good.

One suggestion: we've already got GitPython, and it looks like you can create the remotes using that. It would also save changing directory:

```
import git
...
repo = git.Repo(module_name)
repo.create_remote(name, url)
```

Slightly easier?

**Comment by willrogers** [comment]
GitPython: https://gitpython.readthedocs.io/en/stable/index.html

**Comment by skyfrench** [comment]
Sounds good. I've replaced with calls to GitPython as suggested.

File: dls_ade/dls_checkout_module.py [hunk 1]
```diff
@@ -66,6 +69,44 @@ def check_technical_area(area, module):
...
+                splitline = line.replace('\n', '').split(' ')
+
+                if len(splitline) == 2:
+
+                    remote_name = splitline[0]
+                    remote_url = splitline[1]
+
+                    remotes[remote_name] = remote_url
+
+        for name, url in remotes.items():
```
Comment (GDYendell) [inline dls_ade/dls_checkout_module.py]: Could this for loop go into GitoliteServer with a virtual method on GitServer callled add_remotes (e.g.). That would keep usage of git and gitolite away from the main script.

File: dls_ade/dls_checkout_module.py [hunk 2]
```diff
@@ -66,6 +69,44 @@ def check_technical_area(area, module):
         raise Exception("Missing Technical Area under Beamline")
 
 
+def setup_remotes(module_name):
```
Comment (GDYendell) [inline dls_ade/dls_checkout_module.py]: Could this whole method go in vcs_git.

File: dls_ade/dls_checkout_module.py [hunk 3]
```diff
@@ -66,6 +69,44 @@ def check_technical_area(area, module):
...
+            if name != 'gitolite':  # origin -> named 'gitolite' on original clone
+
+                usermsg.info("Adding remote {remote_name} {remote_url}".format(remote_name=name, remote_url=url))
+                repo = git.Repo(module_name)
+                repo.create_remote(name, url)
+
+    else:
+
+        usermsg.info("No remotes defined.")
+
```
Comment (GDYendell) [inline dls_ade/dls_checkout_module.py]: A few extra and missing lines and a couple of lines too long, as discussed.

Comment (GDYendell) [reply dls_ade/dls_checkout_module.py]: https://www.python.org/dev/peps/pep-0008/?#blank-lines

**Comment by skyfrench** [comment]
Updated with Gary's suggested improvements.

**Comment by willrogers** [comment]
Another idea: you could give a keyword argument with a default to `get_clone_repo()` in `gitserver.py`, then call the super method with `gitolite` in `gitoliteserver.py`:

```
    def get_clone_repo(self, server_repo_path, local_repo_path, origin='origin'):

        repo = git.Repo.clone_from(
            os.path.join(self.clone_url,
                         self.get_clone_path(server_repo_path)),
            os.path.join("./", local_repo_path),
            origin=origin)

        return repo
```

and

```
    def get_clone_repo(self, server_repo_path, local_repo_path):
        super(GitoliteServer, self).get_clone_repo(server_repo_path, local_repo_path, 'gitolite')
```

Does that work? And is it actually better?

**Comment by willrogers** [comment]
Oh no I meant to approve it but I don't seem to have.

**Comment by skyfrench** [comment]
Hopefully final modifications (bug fix spotted by Gary, Will's refactoring suggestion).

**Comment by skyfrench** [comment]
I've added a couple of unittests as required.

**Comment by GDYendell** [comment]
I think this is probably OK to go in now. Would anyone take offence at doing a "Squash and Merge" for this?

**Comment by willrogers** [comment]
I changed a log message and actually added the `.gitremotes` file to this branch.

Merge?

**Pull request merged**